domain/operation/state/task.go

```diff
@@ -0,0 +1,278 @@
+// Copyright 2025 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package state
+
+import (
+	"context"
+	"database/sql"
+
+	"github.com/canonical/sqlair"
+
+	"github.com/juju/collections/transform"
+	"github.com/juju/juju/domain/operation"
+	operationerrors "github.com/juju/juju/domain/operation/errors"
+	"github.com/juju/juju/internal/errors"
+	internaluuid "github.com/juju/juju/internal/uuid"
+)
+
+// GetTask returns the task identified by its ID.
+// It returns the task as well as the path to its output in the object store,
+// if any. It's up to the caller to retrieve the actual output from the object
+// store.
+//
+// The following errors may be returned:
+// - [operationerrors.TaskNotFound] when the task does not exists.
+func (s *State) GetTask(ctx context.Context, taskID string) (operation.Task, *string, error) {
+	db, err := s.DB(ctx)
+	if err != nil {
+		return operation.Task{}, nil, errors.Capture(err)
+	}
+
+	var (
+		result     operation.Task
+		outputPath *string
+	)
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		var err error
+
+		result, outputPath, err = s.getTask(ctx, tx, taskID)
+		return errors.Capture(err)
+	})
+	if err != nil {
+		return operation.Task{}, nil, errors.Errorf("getting action %q: %w", taskID, err)
+	}
+
+	return result, outputPath, nil
+}
+
+// CancelTask attempts to cancel an enqueued task, identified by its
+// ID.
+//
+// The following errors may be returned:
+// - [operationerrors.TaskNotFound] when the task does not exists.
+func (s *State) CancelTask(ctx context.Context, taskID string) (operation.Task, error) {
+	db, err := s.DB(ctx)
+	if err != nil {
+		return operation.Task{}, errors.Capture(err)
+	}
+
+	var result operation.Task
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		// Attempt to cancel the task.
+		err = s.cancelTask(ctx, tx, taskID)
+		if err != nil {
+			return errors.Errorf("cancelling task %q: %w", taskID, err)
+		}
+
+		result, _, err = s.getTask(ctx, tx, taskID)
+		return errors.Capture(err)
+	})
+	if err != nil {
+		return operation.Task{}, errors.Errorf("cancelling action %q: %w", taskID, err)
+	}
+
+	return result, nil
+}
+
+// cancelTask updates a specific task to cancelled status.
+func (s *State) cancelTask(ctx context.Context, tx *sqlair.TX, taskID string) error {
+	taskIDParam := taskIdent{ID: taskID}
+
+	currentStatusQuery := `
+SELECT ots.status_id AS &taskStatus.status_id
+FROM   operation_task_status ots
+JOIN   operation_task ot ON ots.task_uuid = ot.uuid
+WHERE  ot.task_id = $taskIdent.task_id
```

Comment:
todo: rework using `JOIN operation_task_status_value osv ON ots.status_id = osv.id` to return the status not an ID.